File: ios/Classes/Serialization.swift
import Foundation
import KoraIDV

/// Helpers to convert native KoraIDV models to [String: Any] dictionaries
/// for passing across the Flutter MethodChannel.
///
/// Mirrors the RN bridge's handleResult serialization, but returns
/// dictionaries instead of JSON strings (Flutter's StandardMethodCodec
/// handles Map serialization natively).
enum FlutterSerialization {

    // MARK: - ISO 8601 formatter

    private static let iso8601: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    // MARK: - Success result

    static func buildSuccessResult(_ verification: Verification) -> [String: Any] {
        return [
            "type": "success",
            "verification": serializeVerification(verification)
        ]
    }

    // MARK: - Cancelled result

    static func buildCancelledResult() -> [String: Any] {
        return ["type": "cancelled"]
    }

    // MARK: - Verification

    static func serializeVerification(_ v: Verification) -> [String: Any] {
        var map: [String: Any] = [
            "id": v.id,
            "externalId": v.externalId,
            "tenantId": v.tenantId,
            "tier": v.tier,
            "status": v.status.rawValue,
            "createdAt": iso8601.string(from: v.createdAt),
            "updatedAt": iso8601.string(from: v.updatedAt),
        ]

        if let dv = v.documentVerification {
            map["documentVerification"] = serializeDocumentVerification(dv)
        }
        if let fv = v.faceVerification {
            map["faceVerification"] = serializeFaceVerification(fv)
        }
        if let lv = v.livenessVerification {
            map["livenessVerification"] = serializeLivenessVerification(lv)
        }
        if let rs = v.riskSignals {
            map["riskSignals"] = rs.map { serializeRiskSignal($0) }
        }
        if let score = v.riskScore {
            map["riskScore"] = score
        }
        if let completed = v.completedAt {
            map["completedAt"] = iso8601.string(from: completed)
        }

        return map
    }

    // MARK: - Document Verification

    private static func serializeDocumentVerification(_ dv: DocumentVerification) -> [String: Any] {
        var map: [String: Any] = [
            "documentType": dv.documentType,
        ]
        if let v = dv.documentNumber { map["documentNumber"] = v }
        if let v = dv.firstName { map["firstName"] = v }
        if let v = dv.lastName { map["lastName"] = v }
        if let v = dv.dateOfBirth { map["dateOfBirth"] = v }
        if let v = dv.expirationDate { map["expirationDate"] = v }
        if let v = dv.issuingCountry { map["issuingCountry"] = v }
        if let v = dv.mrzValid { map["mrzValid"] = v }
        if let v = dv.authenticityScore { map["authenticityScore"] = v }
        if let v = dv.extractedFields { map["extractedFields"] = v }
        return map
    }

    // MARK: - Face Verification

    private static func serializeFaceVerification(_ fv: FaceVerification) -> [String: Any] {
        return [
            "matchScore": fv.matchScore,
            "matchResult": fv.matchResult,
            "confidence": fv.confidence,
        ]
    }

    // MARK: - Liveness Verification

    private static func serializeLivenessVerification(_ lv: LivenessVerification) -> [String: Any] {
        var map: [String: Any] = [
            "livenessScore": lv.livenessScore,
            "isLive": lv.isLive,
        ]
        if let cr = lv.challengeResults {
            map["challengeResults"] = cr.map { serializeChallengeResult($0) }
        }
        return map
    }

    // MARK: - Challenge Result

    private static func serializeChallengeResult(_ cr: ChallengeResult) -> [String: Any] {
        return [
            "type": cr.type,
            "passed": cr.passed,
            "confidence": cr.confidence,
        ]
    }

    // MARK: - Risk Signal

    private static func serializeRiskSignal(_ rs: RiskSignal) -> [String: Any] {
        return [
            "code": rs.code,
            "severity": rs.severity,
            "message": rs.message,
        ]
    }
}

File: android/src/main/kotlin/com/koraidv/flutter/Serialization.kt
package com.koraidv.flutter

import com.koraidv.sdk.ChallengeResult
import com.koraidv.sdk.DocumentVerification
import com.koraidv.sdk.FaceVerification
import com.koraidv.sdk.LivenessVerification
import com.koraidv.sdk.RiskSignal
import com.koraidv.sdk.Verification
import com.koraidv.sdk.VerificationScores
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Serialization helpers for crossing the Flutter MethodChannel bridge.
 *
 * Converts native SDK models -> HashMap<String, Any?> that Flutter's
 * StandardMethodCodec can transmit to the Dart layer.
 *
 * Unlike the RN bridge which serializes to JSON strings, Flutter's
 * StandardMethodCodec natively supports Maps.
 *
 * Mirrors koraidv-react-native's Serialization.kt adapted for Flutter.
 */
object Serialization {

    private val iso8601Format: SimpleDateFormat
        get() = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }

    /**
     * Build a success result map containing the verification.
     */
    fun buildSuccessResult(verification: Verification): HashMap<String, Any?> {
        return hashMapOf(
            "type" to "success",
            "verification" to serializeVerification(verification)
        )
    }

    /**
     * Build a cancelled result map.
     */
    fun buildCancelledResult(): HashMap<String, Any?> {
        return hashMapOf("type" to "cancelled")
    }

    /**
     * Serialize a Verification to a HashMap.
     */
    fun serializeVerification(v: Verification): HashMap<String, Any?> {
        val map = hashMapOf<String, Any?>(
            "id" to v.id,
            "externalId" to v.externalId,
            "tenantId" to v.tenantId,
            "tier" to v.tier,
            "status" to v.status.value,
            "createdAt" to formatDate(v.createdAt),
            "updatedAt" to formatDate(v.updatedAt),
            "completedAt" to v.completedAt?.let { formatDate(it) },
            "riskScore" to v.riskScore,
        )

        val docVerification = v.documentVerification
        if (docVerification != null) {
            map["documentVerification"] = serializeDocumentVerification(docVerification)
        }
        val faceVerification = v.faceVerification
        if (faceVerification != null) {
            map["faceVerification"] = serializeFaceVerification(faceVerification)
        }
        val livenessVerification = v.livenessVerification
        if (livenessVerification != null) {
            map["livenessVerification"] = serializeLivenessVerification(livenessVerification)
        }
        val scores = v.scores
        if (scores != null) {
            map["scores"] = serializeScores(scores)
        }
        val riskSignals = v.riskSignals
        if (riskSignals != null) {
            map["riskSignals"] = riskSignals.map { serializeRiskSignal(it) }
        }
        return map
    }

    private fun serializeDocumentVerification(dv: DocumentVerification): HashMap<String, Any?> {
        return hashMapOf(
            "documentType" to dv.documentType,
            "documentNumber" to dv.documentNumber,
            "firstName" to dv.firstName,
            "lastName" to dv.lastName,
            "dateOfBirth" to dv.dateOfBirth,
            "expirationDate" to dv.expirationDate,
            "issuingCountry" to dv.issuingCountry,
            "mrzValid" to dv.mrzValid,
            "authenticityScore" to dv.authenticityScore,
            "extractedFields" to dv.extractedFields,
        )
    }

    private fun serializeFaceVerification(fv: FaceVerification): HashMap<String, Any?> {
        return hashMapOf(
            "matchScore" to fv.matchScore,
            "matchResult" to fv.matchResult,
            "confidence" to fv.confidence,
        )
    }

    private fun serializeLivenessVerification(lv: LivenessVerification): HashMap<String, Any?> {
        val map = hashMapOf<String, Any?>(
            "livenessScore" to lv.livenessScore,
            "isLive" to lv.isLive,
        )
        val challengeResults = lv.challengeResults
        if (challengeResults != null) {
            map["challengeResults"] = challengeResults.map { serializeChallengeResult(it) }
        }
        return map
    }

    private fun serializeChallengeResult(cr: ChallengeResult): HashMap<String, Any?> {
        return hashMapOf(
            "type" to cr.type,
            "passed" to cr.passed,
            "confidence" to cr.confidence,
        )
    }

    private fun serializeScores(s: VerificationScores): HashMap<String, Any?> {
        return hashMapOf(
            "documentQuality" to s.documentQuality,
            "documentAuth" to s.documentAuth,
            "faceMatch" to s.faceMatch,
            "liveness" to s.liveness,
            "nameMatch" to s.nameMatch,
            "dataConsistency" to s.dataConsistency,
            "screening" to s.screening,
            "overall" to s.overall,
        )
    }

    private fun serializeRiskSignal(rs: RiskSignal): HashMap<String, Any?> {
        return hashMapOf(
            "code" to rs.code,
            "severity" to rs.severity,
            "message" to rs.message,
        )
    }

    private fun formatDate(date: Date): String {
        return iso8601Format.format(date)
    }
}
